src/GridFS/GridFsUpload.php

```diff
@@ -0,0 +1,164 @@
+     * @param \MongoDB\Collection    $chunksCollection Chunks Collection
+     * @param int32                  $chunkSizeBytes   Size of chunk
+     * @param string                 $filename         Filename to insert
+     * @param array                  $options          File options
+     * @throws InvalidArgumentException
+     */
+    public function __construct(
+        \MongoDB\Collection $filesCollection,
+        \MongoDB\Collection $chunksCollection,
+        $chunkSizeBytes,
+        $filename,
+        array $options=[]
+        )
+    {
+        $this->ctx = hash_init('md5');
+        $uploadDate = time();
+        $objectId = new \MongoDB\BSON\ObjectId();
+        $main_file = [
+            "chunkSize" => $chunkSizeBytes,
+            "filename" => $filename,
+            "uploadDate" => $uploadDate,
+            "_id" => $objectId
+        ];
+
+        $fileOptions = [];
+        if (isset($options['contentType'])) {
+            if (is_string($options['contentType'])) {
+                $fileOptions['contentType'] = $options['contentType'];
+            } else {
+                throw new InvalidArgumentTypeException('"contentType" option', $options['contentType'], 'string');
+            }
+        }
+        if (isset($options['aliases'])) {
+            if (\MongoDB\is_string_array($options['aliases'])) {
+                $fileOptions['aliases'] = $options['aliases'];
+            } else {
+                throw new InvalidArgumentTypeException('"aliases" option', $options['aliases'], 'array of strings');
+            }
+        }
+
+        if (isset($options['metadata'])) {
+            if (is_array($options['metadata']) || is_object($options['metadata'])) {
+                $fileOptions['metadata'] = $options['metadata'];
+            } else {
+                throw new InvalidArgumentTypeException('"metadata" option', $options['metadata'], 'object or array');
+            }
+        }
+
+        $this->file = array_merge($main_file, $fileOptions);
+        parent::__construct($filesCollection, $chunksCollection, $chunkSizeBytes);
+    }
+    /**
+    * Reads data from a stream into GridFS
+    *
+    * @param Stream   $source   Source Stream
+    * @return ObjectId
+    */
+    public function uploadFromStream($source)
+    {
+        while ($data = fread($source, $this->chunkSizeBytes)) {
+            $this->insertChunk($data);
+        }
+        return $this->fileCollectionInsert();
+    }
+    /**
+    * Insert a chunks into GridFS from a string
+    *
+    * @param string  $toWrite   String to upload
+    * @return int
+    */
+    public function insertChunks($toWrite)
+    {
+        rewind($this->buffer);
+        $cached = fread($this->buffer, $this->chunkSizeBytes);
+        $toWrite = $cached . $toWrite;
+        $bytes = str_split($toWrite, $this->chunkSizeBytes);
+
+        fclose($this->buffer);
+        $this->buffer = fopen("php://temp", "w+");
```

Comment:
When do you close `$this->buffer`? It looks like you read it out into a PHP string, concatenate, and then reassign it to a newly opened in-memory stream.

`str_split()` also seems like it'd have overhead, since it's creating an array of chunk-sized strings, which all copy data already found in `$toWrite`. I think we can get by with the loop appending enough bytes from `$toWrite` to hit a chunk (or exhaust `$toWrite` entirely) into the in-memory buffer and then `insertChunk()` on the buffer contents each time we fill it up. Afterwards, we can truncate the buffer back to zero and continue the append process as needed.

I realized we discussed the in-memory buffer quite a bit yesterday, but it occurs to me reading this code that string concatenation may prove simpler. If you want to start with strings, we can always benchmark the in-memory buffer later (just make a ticket so we remember).